### PR TITLE
[FIX-BUG] fix pd-separation decode node free block bug

### PR DESCRIPTION
**Pull request by BJWang-ant**
## Motivation
Fix the issue where, in the pd separation mode, the decode node uses pre-allocated video memory, resulting in duplicate block index problems during block recycling.

## Reproduction environment
B200-16-card
I used deepseekv4 to deploy PD separation, and the startup and testing commands are as follows：
server command：
    p-node：
```
GLOO_SOCKET_IFNAME=eth1 \
    NCCL_SOCKET_IFNAME=eth1 \
    NVSHMEM_IB_GID_INDEX=3 \
    NVSHMEM_IB_TRAFFIC_CLASS=184 \
    NVSHMEM_IB_ENABLE_IBGDA=true \
    MC_SLICE_SIZE=262144 \
    MC_TE_METRIC=true \
    SGLANG_SET_CPU_AFFINITY=1 \
    SGLANG_OPT_SWA_SPLIT_LEAF_ON_INSERT=1 \
    SGLANG_OPT_USE_JIT_NORM=1 \
    SGLANG_OPT_USE_JIT_INDEXER_METADATA=1 \
    SGLANG_OPT_USE_TOPK_V2=1 \
    SGLANG_OPT_SWA_EVICT_DROP_PAGE_MARGIN=1 \
    SGLANG_OPT_USE_FAST_MASK_EP=1 \
    SGLANG_OPT_FIX_MEGA_MOE_MEMORY=1 \
    SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=0 \
    SGLANG_OPT_SWA_RELEASE_LEAF_LOCK_AFTER_WINDOW=1 \
    SGLANG_OPT_USE_DEEPGEMM_MEGA_MOE=1 \
    SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8320 \
    sglang serve \
    --trust-remote-code \
    --model-path /deepseek-ai/DeepSeek-V4-Pro \
    --tp 8 \
    --dp 8 \
    --enable-dp-attention \
    --enable-dp-lm-head \
    --disaggregation-mode prefill \
    --disaggregation-ib-device mlx5_4,mlx5_5,mlx5_6,mlx5_7,mlx5_8,mlx5_9,mlx5_10,mlx5_11 \
    --dist-init-addr 0.0.0.0:5757 \
    --moe-a2a-backend deepep \
    --mem-fraction-static 0.85 \
    --swa-full-tokens-ratio 0.075 \
    --chunked-prefill-size 32768 \
    --enable-prefill-delayer \
    --deepep-config '{"normal_dispatch":{"num_sms":96},"normal_combine":{"num_sms":96}}' \
    --host 0.0.0.0 \
    --port 30001
```
    d-node:
```
GLOO_SOCKET_IFNAME=eth1 \
    NCCL_SOCKET_IFNAME=eth1 \
    NVSHMEM_IB_GID_INDEX=3 \
    NVSHMEM_IB_TRAFFIC_CLASS=184 \
    NVSHMEM_IB_ENABLE_IBGDA=true \
    MC_SLICE_SIZE=262144 \
    MC_TE_METRIC=true \
    SGLANG_SET_CPU_AFFINITY=1 \
    SGLANG_OPT_SWA_SPLIT_LEAF_ON_INSERT=1 \
    SGLANG_OPT_USE_JIT_NORM=1 \
    SGLANG_OPT_USE_JIT_INDEXER_METADATA=1 \
    SGLANG_OPT_USE_TOPK_V2=1 \
    SGLANG_OPT_SWA_EVICT_DROP_PAGE_MARGIN=1 \
    SGLANG_OPT_USE_FAST_MASK_EP=1 \
    SGLANG_OPT_FIX_MEGA_MOE_MEMORY=1 \
    SGLANG_DEEPEP_NUM_MAX_DISPATCH_TOKENS_PER_RANK=0 \
    SGLANG_OPT_SWA_RELEASE_LEAF_LOCK_AFTER_WINDOW=1 \
    SGLANG_OPT_USE_DEEPGEMM_MEGA_MOE=1 \
    SGLANG_OPT_DEEPGEMM_MEGA_MOE_NUM_MAX_TOKENS_PER_RANK=8320 \
    sglang serve \
    --trust-remote-code \
    --model-path /deepseek-ai/DeepSeek-V4-Pro \
    --tp 8 \
    --dp 8 \
    --enable-dp-attention \
    --enable-dp-lm-head \
    --disaggregation-mode decode \
    --disaggregation-ib-device mlx5_4,mlx5_5,mlx5_6,mlx5_7,mlx5_8,mlx5_9,mlx5_10,mlx5_11 \
    --dist-init-addr 0.0.0.0:5757 \
    --moe-a2a-backend deepep \
    --mem-fraction-static 0.85 \
    --max-running-requests 512 \
    --swa-full-tokens-ratio 0.075 \
    --enable-prefill-delayer \
    --deepep-config '{"normal_dispatch":{"num_sms":96},"normal_combine":{"num_sms":96}}' \
    --speculative-algorithm EAGLE \
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 4 \
    --load-balance-method round_robin \
    --host 0.0.0.0 \
    --port 30001
```

benchmark:
```
python3 -m sglang.bench_serving \
      --model /deepseek-ai/DeepSeek-V4-Pro \
      --flush-cache \
      --host 0.0.0.0 \
      --port 8008 \
      --dataset-name "random" \
      --dataset-path "ShareGPT_V3_unfiltered_cleaned_split.json" \
      --num-prompts "2048" \
      --random-range-ratio "1" \
      --random-input-len "4096" \
      --random-output-len "1024" \
      --max-concurrency "256" \
```

then, i get this error:
image

## Modifications
During the process of block recovery when there are situations of occupation, eviction and request for termination, an operation to remove duplicate block indices is added.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** — add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** — `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/mem_cache/allocator.py [hunk 1]
```diff
@@ -503,6 +506,7 @@ def free(self, free_index: torch.Tensor):
             free_page_indices = torch.unique(free_index // self.page_size)
             if self.need_sort:
                 self.release_pages = torch.cat((free_page_indices, self.release_pages))
+                self.release_pages = torch.unique(self.release_pages)
             else:
                 self.free_pages = torch.cat((free_page_indices, self.free_pages))
```
[the remark this answers is not in the document]
Comment (BJWang-ant) [reply python/sglang/srt/mem_cache/allocator.py:511]: I have completed the test and the performance will not be affected.

File: python/sglang/srt/mem_cache/allocator.py [hunk 2]
```diff
@@ -86,6 +88,7 @@ def free_group_end(self):
     def merge_and_sort_free(self):
         if len(self.release_pages) > 0:
             self.free_pages = torch.cat((self.free_pages, self.release_pages))
+            self.free_pages = torch.unique(self.free_pages)
             self.free_pages, _ = torch.sort(self.free_pages)
```
[the remark this answers is not in the document]
Comment (BJWang-ant) [reply python/sglang/srt/mem_cache/allocator.py:92]: I have completed the test and the performance will not be affected.

**Comment by BJWang-ant** [comment]
I test gsm8k datasets acc.
image

**Comment by BJWang-ant** [comment]
@fzyzcjy @Fridge003 Hi, could you please review the code for me?

**Comment by ispobock** [comment]
@BJWang-ant Could you try https://github.com/sgl-project/sglang/pull/25805?

**Comment by BJWang-ant** [comment]
> @BJWang-ant Could you try #25805?

OK, i will try
